include/zephyr/usb/class/usbd_hid.h

```diff
@@ -0,0 +1,207 @@
+ *         +---------------------+       |
+ *         |                     |       |
+ *         |      HID Device     |       |
+ *         |     "bottom half"   |<------+
+ *         |    USB HID class    |
+ *         |    implementation   |
+ *         |                     |
+ *         |                     |
+ *         +---------------------+
+ *                    ^
+ *                    v
+ *          +--------------------+
+ *          |                    |
+ *          |     USB Device     |
+ *          |       Support      |
+ *          |                    |
+ *          +--------------------+
+ */
+
+/**
+ * @brief HID device user callbacks
+ *
+ * Each device depends on a user part that handles feature, input, and output
+ * report processing according to the device functionality described by the
+ * report descriptor. Which callbacks must be implemented depends on the device
+ * functionality. The USB device part of the HID device, cannot interpret
+ * device specific report descriptor and only handles USB specific parts,
+ * transfers and validation of requests, all reports are opaque to it.
+ * Callbacks are called from the USB device stack thread and must not block.
+ */
+struct hid_device_ops {
+	/**
+	 * The interface ready callback is called with the ready argument set
+	 * to true when the corresponding interface is part of the active
+	 * configuration and the application can e.g. begin submitting input
+	 * reports, and with the argument set to false when the interface is no
+	 * longer active. This callback is optional.
+	 */
+	void (*iface_ready)(const struct device *dev, const bool ready);
+
+	/**
+	 * This callback is called for the HID Get Report request to get a
+	 * feature, input, or output report, which is specified by the argument
+	 * type. If there is no report ID in the report descriptor, the id
+	 * argument is zero. The callback implementation must check the
+	 * arguments, such as whether the report type is supported, and return
+	 * a nonzero value to indicate an unsupported type or an error.
+	 */
+	int (*get_report)(const struct device *dev,
+			  const uint8_t type, const uint8_t id,
+			  const uint16_t len, uint8_t *const buf);
+
+	/**
+	 * This callback is called for the HID Set Report request to set a
+	 * feature, input, or output report, which is specified by the argument
+	 * type. If there is no report ID in the report descriptor, the id
+	 * argument is zero. The callback implementation must check the
+	 * arguments, such as whether the report type is supported, and return
+	 * a nonzero value to indicate an unsupported type or an error.
+	 */
+	int (*set_report)(const struct device *dev,
+			  const uint8_t type, const uint8_t id,
+			  const uint16_t len, const uint8_t *const buf);
+
+	/**
+	 * Notification to limit intput report frequency.
+	 * The application should mute an input report submission until a new
+	 * event occurs or until the time specified by the duration value has
```

Comment:
Two things here:
1. It would be nice to mention that the duration is expressed in milliseconds (as opposed to the raw USB value which has 4ms resolution).
2. It would be nice to group the USB-only part of the HID API (set_idle, get_idle, sof) separately, even a prefix would be nice, just as an indication of transport dependent support for these.

Reply:
> Two things here:
> 
>     1. It would be nice to mention that the duration is expressed in milliseconds (as opposed to the raw USB value which has 4ms resolution).

I removed */4U, it is 4 millisecond resolution now.

>     2. It would be nice to group the USB-only part of the HID API (set_idle, get_idle, sof) separately, even a prefix would be nice, just as an indication of transport dependent support for these.

It is USB-only part, it is prefixed with `usbd_`. However it evolves, there is no reason to add another prefix to anything. A callback is either optional or mandatory, it depends on the device. Adding prefixes does not change that.

Reply:
> I removed */4U, it is 4 millisecond resolution now.

I actually preferred having ms resolution, but I'm also [vehemently against](https://github.com/UltimateHackingKeyboard/firmware/issues/390#issuecomment-991797166) the whole idle logic, so as you see fit.

> It is USB-only part, it is prefixed with usbd_.

The header is prefixed, but the API itself (types and methods) isn't, and also the explaining comment hinted at that this is supposed to be transport-independent. Moving on:

How about a suspend - resume callback, I'd see it as more relevant than having a sof callback? On the other hand these events are best handled on a whole device and not on an interface level.

Reply:
> > I removed */4U, it is 4 millisecond resolution now.
> 
> I actually preferred having ms resolution, but I'm also [vehemently against](https://github.com/UltimateHackingKeyboard/firmware/issues/390#issuecomment-991797166) the whole idle logic, so as you see fit.

Changed it back and added comments about duration time resolution.
IMHO Get/Set Idle is an accident. Bluetooth was right to drop it.

> How about a suspend - resume callback, I'd see it as more relevant than having a sof callback? On the other hand these events are best handled on a whole device and not on an interface level.

There is common notification mechanism for the bus events. I also tested HID with suspend/rwup-requests (L2->L0), there will be a follow up PR to enable it for all functions. Also note that this API is experimental; if any limitations become apparent, we can fix them quickly.